### PR TITLE
Add support for loading external GATT specs + parsers from distinct bundles

### PR DESCRIPTION
**Pull request by xrucka**
This commit consists of 2 changes:

1. Allow subclassing FieldHolder (alternativ iders without package fqdn conflicts.
2. Add support for loading GATT xml specs not only from resources allready within classpath, but via explicit call as well. This allows for external bundles to load custom XML's bundled within the extension jar.

Merge in cojunction with [eclipse-smarthome-bluetooth-binding/#54](https://github.com/sputnikdev/eclipse-smarthome-bluetooth-binding/pull/54)

### CODE REVIEW

File: src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java [hunk 1]
```diff
@@ -77,8 +75,9 @@
 
     private static final FilenameFilter XML_FILE_FILTER = (dir, name) -> name.toLowerCase().endsWith(".xml");
```
Comment (vkolotov) [inline src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java]: I think we no longer need it. It was used to load files from classpath, but we no longer do it. Some legacy code that should be removed...

File: src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java [hunk 2]
```diff
@@ -221,6 +226,54 @@ public void loadExtensionsFromFolder(String path) {
         readCharacteristics(getFilesFromFolder(characteristicsFolderName));
     }
 
+    private static URL specResourceURL(URL original, String characteristicType) throws MalformedURLException {
```
Comment (vkolotov) [inline src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java]: It is looking good, but I'd prefer to keep Java code convention to use 'verbs' for methods that do something, e.g. `getSpecResourceURL`

File: src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java [hunk 3]
```diff
@@ -221,6 +226,54 @@ public void loadExtensionsFromFolder(String path) {
...
+        String oldfile = original.getFile();
```
Comment (vkolotov) [inline src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java]: Same thing as above, I'd prefer to use camel case names, e.g. `oldFile`
There are some other instances below as well

File: src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java [hunk 4]
```diff
@@ -221,6 +226,54 @@ public void loadExtensionsFromFolder(String path) {
...
+        return new URL(
+            original.getProtocol(),
+            original.getHost(), 
+            original.getPort(),
+            newfile
+        );
+    }
+
+    private Map<String, URL> catalogToURLs(URL serviceRegistry, Map<String, String> xmlEntry) {
+        Map<String, URL> processed = new HashMap();
```
Comment (vkolotov) [inline src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java]: Looks like you are using Eclipse, it does not complain about type casting in that case, however IDEA does :)

effectively Map<String, URL> is not compatible with HashMap, but the following is:
```
Map<String, URL> processed = new HashMap<String, URL>();
```
or short:

```
Map<String, URL> processed = new HashMap<>();
```

File: src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java [hunk 5]
```diff
@@ -221,6 +226,54 @@ public void loadExtensionsFromFolder(String path) {
...
+        Map<String, String> loadedServices = readRegistryFromCatalogResource(servicesResource);
+        logger.info("Loaded {} GATT specifications from resource {}", loadedServices.size(), servicesResource);
+        Map<String, URL> loadedServicesRegistry = catalogToURLs(servicesResource, loadedServices);
+
+        Map<String, String> loadedCharacteristics = readRegistryFromCatalogResource(characteristicsResource);
+        logger.info("Loaded {} GATT specifications from resource {}", loadedCharacteristics.size(), characteristicsResource);
+        Map<String, URL> loadedCharacteristicsRegistry = catalogToURLs(characteristicsResource, loadedCharacteristics);
+
+        Map<String, String> loadedTypeRegistry = loadedCharacteristics.entrySet().stream()
+            .collect(Collectors.toMap((v) -> v.getValue(), (v) -> v.getKey()));
```
Comment (vkolotov) [inline src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java]: Can be slightly simplified as per this:

```
Collectors.toMap(Map.Entry::getValue, Map.Entry::getKey)
```

File: src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java (hunk 3 above)
Comment (xrucka) [reply src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java]: ok, renamed the locals to better reflect what they hold

File: src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java (hunk 4 above)
Comment (xrucka) [reply src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java]: Just plain vim. Rewrote & will try to remember.

File: src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java (hunk 5 above)
Comment (xrucka) [reply src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java]: ok

File: src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java (hunk 1 above)
Comment (xrucka) [reply src/main/java/org/sputnikdev/bluetooth/gattparser/spec/BluetoothGattSpecificationReader.java]: ok, removed all unused code I've found

**Pull request merged**